{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.optim as optim\n",
    "import torch.nn.functional as F\n",
    "\n",
    "import random\n",
    "import multiprocessing\n",
    "from collections import OrderedDict, defaultdict\n",
    "from itertools import product\n",
    "from copy import deepcopy, copy\n",
    "from namegenerator import NameGenerator"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO:\n",
    "    - Add Bias term\n",
    "    - Custom weights\n",
    "    - Drop out connections (set weight to 0)\n",
    "    - Custom activation per node\n",
    "    \n",
    "    - Add function that resets stagnation for all species\\\n",
    "    - Network Visualisation\n",
    "    - Species Visualisation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Activation functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def leaky_relu(x):\n",
    "    return F.leaky_relu(x)\n",
    "\n",
    "def tanh(x):\n",
    "    return F.Tanh(x)\n",
    "\n",
    "def relu(x):\n",
    "    return F.relu(x)\n",
    "\n",
    "def sigmoid(x):\n",
    "    return F.Sigmoid(x)\n",
    "\n",
    "string_to_activation = {\n",
    "    'leaky_relu' : leaky_relu,\n",
    "    'relu' : relu,\n",
    "    'sigmoid' : sigmoid,\n",
    "    'tanh' : tanh\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Model(nn.Module):\n",
    "    def __init__(self,layer_sizes):\n",
    "        super(Model, self).__init__()\n",
    "        layers = OrderedDict()\n",
    "        \n",
    "        previous_layer_size = layer_sizes[0]\n",
    "        for idx, current_layer_size in enumerate(layer_sizes[1:]):\n",
    "            layers[str(idx)] = nn.Linear(previous_layer_size, current_layer_size)\n",
    "            previous_layer_size = current_layer_size\n",
    "            \n",
    "        self.layers = nn.Sequential(layers)\n",
    "        \n",
    "    def forward(self, x):\n",
    "        return self.model(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "firstname_generator = NameGenerator('names.csv', 3, 12)\n",
    "new_individual_name = firstname_generator.generate_name()\n",
    "previous_names = []\n",
    "surname_generator = NameGenerator('surnames.csv', 3, 12)\n",
    "new_specie_name = surname_generator.generate_name()\n",
    "previous_surnames = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Genotype(object):\n",
    "    def __init__(self, inputs = 144, \n",
    "                 outputs = 12, \n",
    "                 nonlinearities = ['relu','sigmoid','tanh'],\n",
    "                 topology = None,\n",
    "                 feedforward = True,\n",
    "                 max_depth=None,\n",
    "                 max_nodes = float('inf'),\n",
    "                 initial_weight_stdev = 2.0,\n",
    "                 bias_as_node = False,\n",
    "                 p_add_neuron = 0.03, \n",
    "                 p_add_connection = 0.3, \n",
    "                 p_mutate_weight = 0.8,\n",
    "                 p_reset_weight = 0.1,\n",
    "                 p_reenable_connection = 0.01,\n",
    "                 p_disable_connection = 0.01, \n",
    "                 p_reenable_parent = 0.25, \n",
    "                 p_mutate_bias = 0.2,\n",
    "                 p_mutate_type = 0.2,\n",
    "                 stdev_mutate_weight = 1.5,\n",
    "                 stdev_mutate_bias = 0.5,\n",
    "                 weight_range = (-50.,50.),\n",
    "                 distance_excess_weight = 1.0, \n",
    "                 distance_disjoint_weight = 1.0, \n",
    "                 distance_weight = 0.4):\n",
    "        \n",
    "        self.name = next(new_individual_name)\n",
    "        self.specie = None\n",
    "        \n",
    "        self.inputs = inputs\n",
    "        self.outputs = outputs\n",
    "        self.nonlinearities = nonlinearities\n",
    "        self.feedforward = feedforward\n",
    "        self.bias_as_node = bias_as_node\n",
    "        \n",
    "        self.max_depth = max_depth\n",
    "        self.max_nodes = max_nodes\n",
    "        \n",
    "        self.initial_weight_stdev = initial_weight_stdev\n",
    "        self.stdev_mutate_weight = stdev_mutate_weight\n",
    "        self.stdev_mutate_bias = stdev_mutate_bias\n",
    "        self.weight_range = weight_range\n",
    "        \n",
    "        # Mutation Probabilities\n",
    "        self.p_add_neuron = p_add_neuron\n",
    "        self.p_add_connection = p_add_connection\n",
    "        self.p_mutate_weight = p_mutate_weight\n",
    "        self.p_reset_weight = p_reset_weight\n",
    "        self.p_reenable_connection = p_reenable_connection\n",
    "        self.p_disable_connection = p_disable_connection\n",
    "        self.p_reenable_parent = p_reenable_parent\n",
    "        self.p_mutate_bias = p_mutate_bias\n",
    "        self.p_mutate_type = p_mutate_type\n",
    "        \n",
    "        # Distance weights\n",
    "        self.distance_excess_weight = distance_excess_weight\n",
    "        self.distance_disjoint_weight = distance_disjoint_weight\n",
    "        self.distance_weight = distance_weight\n",
    "        \n",
    "        # Tuples of: id, non_linearity, layer\n",
    "        self.neuron_genes = []\n",
    "        # Tuples of: innovation number, input, output, weight, enabled\n",
    "        self.connection_genes = {}\n",
    "        # Hyperparameter genes\n",
    "        self.hyperparameter_genes = []\n",
    "        \n",
    "        self._initialise_topology(topology)\n",
    "    \n",
    "    def change_specie(self,specie):\n",
    "        self.specie = specie\n",
    "        \n",
    "    def _initialise_topology(self, topology):\n",
    "        if topology is None:\n",
    "            # Initialise inputs\n",
    "            for i in range(self.inputs):\n",
    "                self.neuron_genes.append([i * 2048, random.choice(self.nonlinearities),1.0,0])\n",
    "\n",
    "            # Initialise outputs\n",
    "            for i in range(self.outputs):\n",
    "                self.neuron_genes.append([(self.inputs + i) * 2048, random.choice(self.nonlinearities),1.0,0])\n",
    "\n",
    "            # Initialise connections\n",
    "            innovation_number = 0\n",
    "            for i in range(self.inputs):\n",
    "                for j in range(self.inputs,self.inputs + self.outputs):\n",
    "                    weight = self._initialise_weight(self.inputs,self.outputs)\n",
    "                    self.connection_genes[(i,j)] = [innovation_number, i, j, weight ,True]\n",
    "                    innovation_number += 1\n",
    "        else:\n",
    "            raise NotImplementedError\n",
    "\n",
    "                \n",
    "    def _initialise_weight(self, input_neurons, output_neurons):\n",
    "        weight = np.random.rand()*np.sqrt(1/(input_neurons + output_neurons))\n",
    "        return weight\n",
    "        \n",
    "    def recombinate(self, other):\n",
    "        child = deepcopy(self)\n",
    "        child.neuron_genes = []\n",
    "        child.connection_genes = {}\n",
    "        \n",
    "        max_neurons = max(len(self.neuron_genes), len(other.neuron_genes))\n",
    "        min_neurons = min(len(self.neuron_genes), len(other.neuron_genes))\n",
    "        \n",
    "        for i in range(max_neurons):\n",
    "            neuron_gene = None\n",
    "            if i < min_neurons:\n",
    "                neuron_gene = random.choice((self.neuron_genes[i], other.neuron_genes[i]))\n",
    "            else:\n",
    "                try:\n",
    "                    neuron_gene = self.neuron_genes[i]\n",
    "                except IndexError:\n",
    "                    neuron_gene = other.neuron_genes[i]\n",
    "            child.neuron_genes.append(deepcopy(neuron_gene))\n",
    "            \n",
    "        self_connections = dict(((c[0], c) for c in self.connection_genes.values()))\n",
    "        other_connections = dict(((c[0], c) for c in other.connection_genes.values()))\n",
    "        max_innovation_number = max(list(self_connections.keys()) + list(other_connections.keys()))\n",
    "        \n",
    "        for i in range(max_innovation_number + 1):\n",
    "            connection_gene = None\n",
    "            if i in self_connections and i in other_connections:\n",
    "                connection_gene = random.choice((self_connections[i],other_connections[i]))\n",
    "                enabled = self_connections[i][4] and other_connections[i][4]\n",
    "            else:\n",
    "                if i in self_connections:\n",
    "                    connection_gene = self_connections[i]\n",
    "                    enabled = connection_gene[4]\n",
    "                elif i in other_connections:\n",
    "                    connection_gene = other_connections[i]\n",
    "                    enabled = connection_gene[4]\n",
    "            if connection_gene is not None:\n",
    "                child.connection_genes[(connection_gene[1],connection_gene[2])] = deepcopy(connection_gene)\n",
    "                child.connection_genes[(connection_gene[1],connection_gene[2])][4] = enabled or np.random.rand() < self.p_reenable_parent\n",
    "\n",
    "            def is_feedforward(item):\n",
    "                ((fr, to), cg) = item\n",
    "                return child.neuron_genes[fr][0] < child.neuron_genes[to][0]\n",
    "\n",
    "            if self.feedforward:\n",
    "                child.connection_genes = dict(filter(is_feedforward, child.connection_genes.items()))\n",
    "            \n",
    "        return child\n",
    "        \n",
    "    def mutate(self, innovations = {}, global_innovation_number = 0):\n",
    "        maximum_innovation_number = max(global_innovation_number, max(cg[0] for cg in self.connection_genes.values()))\n",
    "        # TODO: move to separate functions\n",
    "        if len(self.neuron_genes) < self.max_nodes and np.random.rand() < self.p_add_neuron:\n",
    "            possible_to_split = self.connection_genes.keys()\n",
    "            \n",
    "            if self.max_depth is not None:\n",
    "                possible_to_split = [(fr, to) for (fr, to) in possible_to_split if self.neuron_genes[fr][4] + 1 < self.neuron_genes[to][4]]\n",
    "            \n",
    "            if possible_to_split:\n",
    "\n",
    "                # Choose connection to split\n",
    "                split_neuron = self.connection_genes[random.choice(list(self.connection_genes.keys()))]\n",
    "                # Disable old connection\n",
    "                split_neuron[4] = False\n",
    "\n",
    "                input_neuron, output_neuron, weight = split_neuron[1:4]\n",
    "                neuron_id = (self.neuron_genes[input_neuron][0] + self.neuron_genes[input_neuron][0]) * 0.5\n",
    "                nonlinearity = random.choice(self.nonlinearities)\n",
    "                layer = self.neuron_genes[input_neuron][3] + 1\n",
    "\n",
    "                neuron = [neuron_id, nonlinearity, 1.0, layer]\n",
    "\n",
    "                new_id = len(self.neuron_genes)\n",
    "\n",
    "                self.neuron_genes.append(neuron)\n",
    "                \n",
    "                if (input_neuron, new_id) in innovations:\n",
    "                    innovation_number = innovations[(input_neuron,new_id)]\n",
    "                else:\n",
    "                    maximum_innovation_number += 1\n",
    "                    innovation_number = innovations[(input_neuron,new_id)] = maximum_innovation_number\n",
    "                    \n",
    "                # 1.0 to initialise_weight?\n",
    "                self.connection_genes[(input_neuron, new_id)] = [innovation_number, input_neuron, new_id, 1.0, True]\n",
    "                \n",
    "                if (new_id, output_neuron) in innovations:\n",
    "                    innovation_number = innovations[(new_id, output_neuron)]\n",
    "                else:\n",
    "                    maximum_innovation_number += 1\n",
    "                    innovation_number = innovations[(new_id, output_neuron)] = maximum_innovation_number\n",
    "                    \n",
    "                self.connection_genes[(new_id, output_neuron)] = [innovation_number, new_id, output_neuron, weight, True]\n",
    "\n",
    "        elif np.random.rand() < self.p_add_connection:\n",
    "            potential_connections = product(range(len(self.neuron_genes)),range(self.inputs, len(self.neuron_genes)))\n",
    "            potential_connections = (connection for connection in potential_connections if connection not in self.connection_genes)\n",
    "            \n",
    "            if self.feedforward:\n",
    "                potential_connections = ((f, t) for (f, t) in potential_connections if self.neuron_genes[f][0] < self.neuron_genes[t][0])\n",
    "            \n",
    "            potential_connections = list(potential_connections)\n",
    "            if potential_connections:\n",
    "                (fr, to) = random.choice(potential_connections)\n",
    "                if (fr, to) in innovations:\n",
    "                    innovation = innovations[(fr, to)]\n",
    "                else:\n",
    "                    maximum_innovation_number += 1\n",
    "                    innovation = innovations[(fr, to)] = maximum_innovation_number\n",
    "                # get number of neurons in layers of fr and to\n",
    "                connection_gene = [innovation, fr, to, self._initialise_weight(2,2), True]\n",
    "                self.connection_genes[(fr, to)] = connection_gene\n",
    "        else:\n",
    "            for cg in self.connection_genes.values():\n",
    "                if np.random.rand() < self.p_mutate_weight:\n",
    "                    cg[3] += np.random.normal(0, self.stdev_mutate_weight)\n",
    "                    cg[3] = np.clip(cg[3], self.weight_range[0], self.weight_range[1])\n",
    "                    # clipping?\n",
    "                if np.random.rand() < self.p_reset_weight:\n",
    "                    cg[3] = np.random.normal(0,self.stdev_mutate_weight)\n",
    "                    \n",
    "                # bigger chance to disable in this way\n",
    "                if np.random.rand() < self.p_reenable_connection:\n",
    "                    cg[4] = True\n",
    "                    \n",
    "                if np.random.rand() < self.p_disable_connection:\n",
    "                    cg[4] = False\n",
    "                    \n",
    "            for neuron_gene in self.neuron_genes[self.inputs:]:\n",
    "                if np.random.rand() < self.p_mutate_bias:\n",
    "                    neuron_gene[2] += np.random.normal(0, 1)\n",
    "                    neuron_gene[2] = np.clip(neuron_gene[2], self.weight_range[0], self.weight_range[1])\n",
    "                \n",
    "                if np.random.rand() < self.p_mutate_type:\n",
    "                    neuron_gene[1] = random.choice(self.nonlinearities)   \n",
    "                    \n",
    "        return self\n",
    "        \n",
    "    def distance(self, other):\n",
    "        self_connections = dict(((c[0], c) for c in self.connection_genes.values()))\n",
    "        other_connections = dict(((c[0], c) for c in other.connection_genes.values()))\n",
    "\n",
    "        all_innovations = list(self_connections.keys()) + list(other_connections.keys())\n",
    "\n",
    "        minimum_innovation = min(all_innovations)\n",
    "        \n",
    "        e = 0\n",
    "        d = 0\n",
    "        w = 0.0\n",
    "        m = 0\n",
    "        \n",
    "        for innovation_key in all_innovations:\n",
    "            if innovation_key in self_connections and innovation_key in other_connections:\n",
    "                w += np.abs(self_connections[innovation_key][3] - other_connections[innovation_key][3])\n",
    "                m += 1\n",
    "            elif innovation_key in self_connections or innovation_key in other_connections:\n",
    "                # Disjoint genes\n",
    "                if innovation_key < minimum_innovation:\n",
    "                    d += 1\n",
    "                # Excess genes\n",
    "                else:\n",
    "                    e += 1\n",
    "                    \n",
    "        # Average weight differences of matching genes\n",
    "        w = (w/m) if m>0 else w\n",
    "        \n",
    "        return (self.distance_excess_weight * e +\n",
    "               self.distance_disjoint_weight * d +\n",
    "               self.distance_weight * w)\n",
    "    \n",
    "    def get_network_data(self):\n",
    "        \"\"\" Returns a tuple of (connection_matrix, node_types) \n",
    "            that is reordered by the \"feed-forward order\" of the network,\n",
    "            Such that if feedforward was set to true, the matrix will be\n",
    "            lower-triangular.\n",
    "            The node bias is inserted as \"node 0\", the leftmost column\n",
    "            of the matrix.\n",
    "        \"\"\"\n",
    "        \n",
    "        # Assemble connectivity matrix\n",
    "        cm = np.zeros((len(self.neuron_genes), len(self.neuron_genes)))\n",
    "        cm.fill(np.nan)\n",
    "        \n",
    "        for (_, fr, to, weight, enabled) in self.connection_genes.values():\n",
    "            if enabled:\n",
    "                cm[to, fr] = weight\n",
    "        \n",
    "        # Reorder the nodes/connections\n",
    "#         ff, node_types, bias, response, layer = zip(*self.neuron_genes)\n",
    "        ff, node_types, bias, layer = zip(*self.neuron_genes)\n",
    "        order = [i for _,i in sorted(zip(ff, range(len(ff))))]\n",
    "        cm = cm[:,order][order,:]\n",
    "        node_types = np.array(node_types)[order]\n",
    "        bias = np.array(bias)[order]\n",
    "#         response = np.array(response)[order]\n",
    "        layers = np.array(layer)[order]\n",
    "\n",
    "#         # Then, we multiply all the incoming connection weights by the response\n",
    "        cm *= np.atleast_2d(4.9).T\n",
    "        # Finally, add the bias as incoming weights from node-0\n",
    "        if not self.bias_as_node:\n",
    "            cm = np.hstack( (np.atleast_2d(bias).T, cm) )\n",
    "            cm = np.insert(cm, 0, 0.0, axis=0)\n",
    "            # TODO: this is a bit ugly, we duplicate the first node type for \n",
    "            # bias node. It shouldn't matter though since the bias is used as an input.\n",
    "            node_types = [node_types[0]] + list(node_types)\n",
    "\n",
    "        if self.feedforward and np.triu(np.nan_to_num(cm)).any():\n",
    "            import pprint\n",
    "            pprint.pprint(self.neuron_genes)\n",
    "            pprint.pprint(self.connection_genes)\n",
    "            print(ff)\n",
    "            print(order)\n",
    "            print(np.sign(cm))\n",
    "            raise Exception(\"Network is not feedforward.\")\n",
    "        \n",
    "        return cm, node_types"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Species(object):\n",
    "    def __init__(self, initial_member):\n",
    "        self.name = next(new_specie_name)\n",
    "        self.members = [initial_member]\n",
    "        self.representative = initial_member\n",
    "        self.offspring = 0\n",
    "        self.age = 0\n",
    "        self.average_fitness = 0.\n",
    "        self.max_fitness = 0.\n",
    "        self.max_fitness_previous = 0.0\n",
    "        self.stagnation = 0\n",
    "        self.has_best = False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Population(object):\n",
    "    def __init__(self, genome_factory,\n",
    "                population_size = 100,\n",
    "                elitism = True,\n",
    "                stop_when_solved = False,\n",
    "                tournament_selection_k = 3,\n",
    "                verbose = True,\n",
    "                max_cores = 1,\n",
    "                compatibility_threshold = 3.0,\n",
    "                compatibility_threshold_delta = 0.4,\n",
    "                target_species = 12,\n",
    "                minimum_elitism_size = 5,\n",
    "                young_age = 10,\n",
    "                young_multiplier = 1.2,\n",
    "                old_age = 30,\n",
    "                old_multiplier = 0.2,\n",
    "                stagnation_age = 15,\n",
    "                reset_innovations = False,\n",
    "                survival = 0.2):\n",
    "        \n",
    "        self.genome_factory = genome_factory\n",
    "        self.population_size = population_size\n",
    "        self.elitism = elitism\n",
    "        self.stop_when_solved = stop_when_solved\n",
    "        self.tournament_selection_k = tournament_selection_k\n",
    "        self.verbose = verbose\n",
    "        self.max_cores = max_cores\n",
    "        \n",
    "        cpus = multiprocessing.cpu_count()\n",
    "        use_cores = min(self.max_cores, cpus-1)\n",
    "        if use_cores > 1:\n",
    "            self.pool = multiprocessing.Pool(processes=use_cores, maxtasksperchild=5)\n",
    "        else:\n",
    "            self.pool = None\n",
    "        \n",
    "        self.compatibility_threshold = compatibility_threshold\n",
    "        self.compatibility_threshold_delta = compatibility_threshold_delta\n",
    "        \n",
    "        self.target_species = target_species\n",
    "        self.minimum_elitism_size = minimum_elitism_size\n",
    "        \n",
    "        self.young_age = young_age\n",
    "        self.young_multiplier = young_multiplier\n",
    "        self.old_age = old_age\n",
    "        self.old_multiplier = old_multiplier\n",
    "        \n",
    "        self.stagnation_age = stagnation_age\n",
    "        \n",
    "        self.reset_innovations = reset_innovations\n",
    "        self.survival = survival\n",
    "    \n",
    "    # Peas has this function outside the class\n",
    "    def evaluate_individual(self,item):\n",
    "        (individual, evaluator) = item\n",
    "        if callable(evaluator):\n",
    "            individual.stats = evaluator(individual)\n",
    "        elif hasattr(evaluator, 'evaluate'):\n",
    "            individual.stats = evaluator.evaluate(individual)\n",
    "        else:\n",
    "            raise Exception(\"Evaluator must be a callable or object\" \\\n",
    "                        \"with a callable attribute 'evaluate'.\")\n",
    "        return individual\n",
    "        \n",
    "    def _evaluate_all(self, population, evaluator):\n",
    "        to_eval = [(individual, evaluator) for individual in population]\n",
    "        if self.pool is not None:\n",
    "            population = list(self.pool.map(self.evaluate_indivual, to_eval))\n",
    "        else:\n",
    "            population = list(map(self.evaluate_individual, to_eval))\n",
    "        \n",
    "        return population\n",
    "        \n",
    "    def _reset(self):\n",
    "        self.champions = []\n",
    "        self.generation = 0\n",
    "        self.solved_at = None\n",
    "        self.stats = defaultdict(list)\n",
    "        self.species = []\n",
    "        self.global_innovation_number = 0\n",
    "        self.innovations = {}\n",
    "        self.current_compatibility_threshold = self.compatibility_threshold\n",
    "        \n",
    "    def _find_best(self, population, solution = None):\n",
    "        self.champions.append(max(population, key=lambda individual: individual.stats['fitness']))\n",
    "        \n",
    "        if solution is not None:\n",
    "            if isinstance(solution, (int, float)):\n",
    "                solved = (self.champions[-1].stats['fitness'] >= solution)\n",
    "            elif callable(solution):\n",
    "                solved = solution(self.champions[-1])\n",
    "            elif hasattr(solution, 'solve'):\n",
    "                solved = solution.solve(self.champions[-1])\n",
    "                \n",
    "            if solved and self.solved_at is None:\n",
    "                self.solved_at = self.generation + 1\n",
    "            \n",
    "    @property\n",
    "    def population(self):\n",
    "        for specie in self.species:\n",
    "            for member in specie.members:\n",
    "                yield member\n",
    "    \n",
    "    def _evolve(self, evaluator, solution=None):\n",
    "        population = list(self.population)\n",
    "        \n",
    "        while len(population) < self.population_size:\n",
    "            individual = self.genome_factory()\n",
    "            population.append(individual)        \n",
    "            \n",
    "        population = self._evaluate_all(population, evaluator)\n",
    "        \n",
    "        # Speciation\n",
    "        for specie in self.species:\n",
    "            # Choose random specie representative for distance comparison\n",
    "            specie.representative = random.choice(specie.members)\n",
    "            specie.name = specie.representative.specie\n",
    "            specie.members = []\n",
    "            specie.age += 1\n",
    "            \n",
    "        # Add each individual to a species\n",
    "        for individual in population:\n",
    "            found = False\n",
    "            for specie in self.species:\n",
    "                if individual.distance(specie.representative) <= self.current_compatibility_threshold:\n",
    "                    specie.members.append(individual)\n",
    "                    individual.change_specie(specie.name)\n",
    "                    found = True\n",
    "                    break\n",
    "            if not found:\n",
    "                s = Species(individual)\n",
    "                individual.change_specie(s.name)\n",
    "                self.species.append(s)\n",
    "        \n",
    "        # Remove empty species\n",
    "        self.species = list(filter(lambda s: len(s.members) > 0, self.species))\n",
    "        \n",
    "        # Adjust compatibility threshold\n",
    "        if len(self.species) < self.target_species:\n",
    "            self.current_compatibility_threshold -= self.compatibility_threshold_delta\n",
    "        elif len(self.species) > self.target_species:\n",
    "            self.current_compatibility_threshold += self.compatibility_threshold_delta\n",
    "        \n",
    "        # Find champion and check for solution\n",
    "        self._find_best(population, solution)\n",
    "        \n",
    "        # Recombination\n",
    "        \n",
    "        for specie in self.species:\n",
    "            specie.max_fitness_previous = specie.max_fitness\n",
    "            specie.average_fitness = np.mean([individual.stats['fitness'] for individual in specie.members])\n",
    "            specie.max_fitness = np.max([individual.stats['fitness'] for individual in specie.members])\n",
    "            if specie.max_fitness <= specie.max_fitness_previous:\n",
    "                specie.stagnation += 1\n",
    "            else:\n",
    "                specie.stagnation = 0\n",
    "            specie.has_best = self.champions[-1] in specie.members\n",
    "        \n",
    "        # Keep species that have the best or within stagnation age range\n",
    "        self.species = list(filter(lambda s: s.stagnation < self.stagnation_age or s.has_best, self.species))\n",
    "        \n",
    "        average_fitness = np.array([specie.average_fitness for specie in self.species])\n",
    "        \n",
    "        # Adjust fitness based on age\n",
    "        age = np.array([specie.age for specie in self.species])\n",
    "        for specie in self.species:\n",
    "            if specie.age < self.young_age:\n",
    "                specie.average_fitness *= self.young_multiplier\n",
    "            if specie.age > self.old_age:\n",
    "                specie.average_fitness *= self.old_multiplier\n",
    "                \n",
    "        # Compute offspring size\n",
    "        total_fitness = sum(specie.average_fitness for specie in self.species)\n",
    "        for specie in self.species:\n",
    "            specie.offspring = int(round(self.population_size * specie.average_fitness / total_fitness))\n",
    "            \n",
    "        \n",
    "        \n",
    "        # Remove species without offspring\n",
    "        self.species = list(filter(lambda s: s.offspring > 0, self.species))\n",
    "\n",
    "        for specie in self.species:\n",
    "            specie.members.sort(key=lambda individual: individual.stats['fitness'], reverse = True)\n",
    "            keep = max(1, int(round(len(specie.members)*self.survival)))\n",
    "            pool = specie.members[:keep]\n",
    "            \n",
    "            if self.elitism and len(specie.members) > self.minimum_elitism_size:\n",
    "                specie.members = specie.members[:1]\n",
    "            else:\n",
    "                specie.members = []\n",
    "                \n",
    "            while len(specie.members) < specie.offspring:\n",
    "                k = min(len(pool), self.tournament_selection_k)\n",
    "                p1 = max(random.sample(pool,k), key=lambda individual: individual.stats['fitness'])\n",
    "                p2 = max(random.sample(pool,k), key=lambda individual: individual.stats['fitness'])\n",
    "                \n",
    "                child = p1.recombinate(p2)\n",
    "                child.mutate(innovations=self.innovations, global_innovation_number = self.global_innovation_number)\n",
    "                specie.members.append(child)\n",
    "                \n",
    "        if self.innovations:\n",
    "            self.global_innovation_number = max(self.innovations.values())\n",
    "            \n",
    "        self._gather_stats(population)\n",
    "            \n",
    "    def epoch(self, evaluator, generations, solution=None, reset=True, callback= None):\n",
    "        if reset:\n",
    "            self._reset()\n",
    "            \n",
    "        for i in range(generations):\n",
    "            self._evolve(evaluator, solution)\n",
    "            self.generation += 1\n",
    "            \n",
    "            if self.verbose:\n",
    "                self._status_report()\n",
    "                \n",
    "            if callback is not None:\n",
    "                callback(self)\n",
    "            \n",
    "            if self.solved_at is not None and self.stop_when_solved:\n",
    "                break\n",
    "        \n",
    "        return {'stats': self.stats, 'champions': self.champions}\n",
    "    \n",
    "    def _gather_stats(self, population):\n",
    "        for key in population[0].stats:\n",
    "            self.stats[key+'_avg'].append(np.mean([individual.stats[key] for individual in population]))\n",
    "            self.stats[key+'_max'].append(np.max([individual.stats[key] for individual in population]))\n",
    "            self.stats[key+'_min'].append(np.min([individual.stats[key] for individual in population]))\n",
    "        self.stats['solved'].append( self.solved_at is not None )\n",
    "    \n",
    "    def _status_report(self):\n",
    "        print(\"\\n****** Running Generation %d ******\" % self.generation)\n",
    "        fitness_list = np.array([i.stats['fitness'] for i in self.population])\n",
    "        number_neurons = len(self.champions[-1].neuron_genes)\n",
    "        number_enabled_connections = np.sum([1 for conn in self.champions[-1].connection_genes.values() if conn[4]])\n",
    "        print(\"Population's average fitness: %.5f stdev: %.5f\" % (np.average(fitness_list), np.std(fitness_list)))\n",
    "        print(\"Best individual: %s %s\" % (self.champions[-1].specie, self.champions[-1].name))\n",
    "        print(\"Best fitness: %.2f - #neurons: %i - #enabled connections: %i\" % (self.champions[-1].stats['fitness'],number_neurons,number_enabled_connections))\n",
    "        print(\"Population of %i members in %i species:\" % (len(list(self.population)), len(self.species)))\n",
    "        print(\"Species         age    size    fitness    stag\")\n",
    "        print(\"============    ===    ====    =======    ====\")\n",
    "        for specie in self.species:\n",
    "            print(\"{: >12}    {: >3}    {: >4}    {:.5f}    {: >4}\".format(specie.name,specie.age,len(specie.members),specie.max_fitness,specie.stagnation))\n",
    "        print(\"Solved: %s\" % (self.solved_at))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Package with some classes to simulate neural nets.\n",
    "\"\"\"\n",
    "\n",
    "### IMPORTS ###\n",
    "\n",
    "import sys\n",
    "import numpy as np\n",
    "np.seterr(over='ignore', divide='raise')\n",
    "\n",
    "# Libraries\n",
    "\n",
    "# Local\n",
    "\n",
    "\n",
    "# Shortcuts\n",
    "\n",
    "inf = float('inf')\n",
    "sqrt_two_pi = np.sqrt(np.pi * 2)\n",
    "\n",
    "### FUNCTIONS ###\n",
    "\n",
    "# Node functions\n",
    "def ident(x):\n",
    "    return x\n",
    "\n",
    "def bound(x, clip=(-1.0, 1.0)):\n",
    "    return np.clip(x, *clip)\n",
    "\n",
    "def gauss(x):\n",
    "    \"\"\" Returns the pdf of a gaussian.\n",
    "    \"\"\"\n",
    "    return np.exp(-x ** 2 / 2.0) / sqrt_two_pi\n",
    "    \n",
    "def sigmoid(x):\n",
    "    \"\"\" Sigmoid function. \n",
    "    \"\"\"\n",
    "    return 1 / (1 + np.exp(-x))\n",
    "\n",
    "def sigmoid2(x):\n",
    "    \"\"\" Sigmoid function. \n",
    "    \"\"\"\n",
    "    return 1 / (1 + np.exp(-4.9*x))\n",
    "\n",
    "def abs(x):\n",
    "    return np.abs(x)\n",
    "\n",
    "def sin(x):\n",
    "    return np.sin(x)\n",
    "\n",
    "def tanh(x):\n",
    "    return np.tanh(x)\n",
    "\n",
    "def summed(fn):\n",
    "    return lambda x: fn(sum(x))\n",
    "\n",
    "def relu(x):\n",
    "    return x * (x>0)\n",
    "\n",
    "### CONSTANTS ###\n",
    "\n",
    "SIMPLE_NODE_FUNCS = {\n",
    "    'sin': np.sin,\n",
    "    'abs': np.abs,\n",
    "    'ident': ident,\n",
    "    'linear': ident,\n",
    "    'bound': bound,\n",
    "    'gauss': gauss,\n",
    "    'sigmoid': sigmoid,\n",
    "    'sigmoid2': sigmoid2,\n",
    "    'exp': sigmoid,\n",
    "    'tanh': tanh,\n",
    "    'relu': relu,\n",
    "    None : ident\n",
    "}\n",
    "\n",
    "def rbfgauss(x):\n",
    "    return np.exp(-(x ** 2).sum() / 2.0) / sqrt_two_pi\n",
    "\n",
    "def rbfwavelet(x):\n",
    "    return np.exp(-(x ** 2).sum() / ( 2* 0.5**2 )) * np.sin(2 * np.pi * x[0])\n",
    "\n",
    "COMPLEX_NODE_FUNCS = {\n",
    "    'rbfgauss': rbfgauss,\n",
    "    'rbfwavelet': rbfwavelet\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "### CLASSES ### \n",
    "\n",
    "class NeuralNetwork(object):\n",
    "    \"\"\" A neural network. Can have recursive connections.\n",
    "    \"\"\"\n",
    "    \n",
    "    def from_matrix(self, matrix, node_types=['sigmoid']):\n",
    "        \"\"\" Constructs a network from a weight matrix. \n",
    "        \"\"\"\n",
    "        # Initialize net\n",
    "        self.original_shape = matrix.shape[:matrix.ndim//2]\n",
    "        # If the connectivity matrix is given as a hypercube, squash it down to 2D\n",
    "        n_nodes = np.prod(self.original_shape)\n",
    "        self.cm  = matrix.reshape((n_nodes,n_nodes))\n",
    "        self.node_types = node_types\n",
    "        if len(self.node_types) == 1:\n",
    "            self.node_types *= n_nodes\n",
    "        self.act = np.zeros(self.cm.shape[0])\n",
    "        self.optimize()\n",
    "        return self\n",
    "        \n",
    "    def from_neatchromosome(self, chromosome):\n",
    "        \"\"\" Construct a network from a Chromosome instance, from\n",
    "            the neat-python package. This is a connection-list\n",
    "            representation.\n",
    "        \"\"\"\n",
    "        # TODO Deprecate the neat-python compatibility\n",
    "        # Typecheck\n",
    "        import neat.chromosome\n",
    "        \n",
    "        if not isinstance(chromosome, neat.chromosome.Chromosome):\n",
    "            raise Exception(\"Input should be a NEAT chromosome, is %r.\" % (chromosome))\n",
    "        # Sort nodes: BIAS, INPUT, HIDDEN, OUTPUT, with HIDDEN sorted by feed-forward.\n",
    "        nodes = dict((n.id, n) for n in chromosome.neuron_genes)\n",
    "        node_order = ['bias']\n",
    "        node_order += [n.id for n in filter(lambda n: n.type == 'INPUT', nodes.values())]\n",
    "        if isinstance(chromosome, neat.chromosome.FFChromosome):\n",
    "            node_order += chromosome.node_order\n",
    "        else:\n",
    "            node_order += [n.id for n in filter(lambda n: n.type == 'HIDDEN', nodes.values())]\n",
    "        node_order += [n.id for n in filter(lambda n: n.type == 'OUTPUT', nodes.values())]\n",
    "        # Construct object\n",
    "        self.cm = np.zeros((len(node_order), len(node_order)))\n",
    "        # Add bias connections\n",
    "        for id, node in nodes.items():\n",
    "            self.cm[node_order.index(id), 0] = node.bias\n",
    "            self.cm[node_order.index(id), 1:] = node.response\n",
    "        # Add the connections\n",
    "        for conn in chromosome.connection_genes:\n",
    "            if conn.enabled:\n",
    "                to = node_order.index(conn.outnodeid)\n",
    "                fr = node_order.index(conn.innodeid)\n",
    "                # dir(conn.weight)\n",
    "                self.cm[to, fr] *= conn.weight\n",
    "        # Verify actual feed forward\n",
    "        if isinstance(chromosome, neat.chromosome.FFChromosome):\n",
    "            if np.triu(self.cm).any():\n",
    "                raise Exception(\"NEAT Chromosome does not describe feedforward network.\")\n",
    "        node_order.remove('bias')\n",
    "        self.node_types = [nodes[i].activation_type for i in node_order]\n",
    "        self.node_types = ['ident'] + self.node_types\n",
    "        self.act = np.zeros(self.cm.shape[0])\n",
    "        self.optimize()\n",
    "        return self\n",
    "\n",
    "    def optimize(self):\n",
    "        # If all nodes are simple nodes\n",
    "        if all(fn in SIMPLE_NODE_FUNCS for fn in self.node_types):\n",
    "            # Simply always sum the node inputs, this is faster\n",
    "            self.sum_all_node_inputs = True\n",
    "            self.cm = np.nan_to_num(self.cm)\n",
    "            # If all nodes are identical types\n",
    "            if all(fn == self.node_types[0] for fn in self.node_types):\n",
    "                self.all_nodes_same_function = True\n",
    "            self.node_types = [SIMPLE_NODE_FUNCS[fn] for fn in self.node_types]\n",
    "        else:\n",
    "            nt = []\n",
    "            for fn in self.node_types:\n",
    "                if fn in SIMPLE_NODE_FUNCS:\n",
    "                    # Substitute the function(x) for function(sum(x))\n",
    "                    nt.append(summed(SIMPLE_NODE_FUNCS[fn]))\n",
    "                else:\n",
    "                    nt.append(COMPLEX_NODE_FUNCS[fn])\n",
    "            self.node_types = nt\n",
    "\n",
    "    \n",
    "    def __init__(self, source=None):\n",
    "        # Set instance vars\n",
    "        self.feedforward    = False\n",
    "        self.sandwich       = False   \n",
    "        self.cm             = None\n",
    "        self.node_types     = None\n",
    "        self.original_shape = None\n",
    "        self.sum_all_node_inputs = False\n",
    "        self.all_nodes_same_function = False\n",
    "        \n",
    "        if source is not None:\n",
    "            try:\n",
    "                self.from_matrix(*source.get_network_data())\n",
    "                if hasattr(source, 'feedforward') and source.feedforward:\n",
    "                    self.make_feedforward()\n",
    "            except AttributeError as e:\n",
    "                print(e)\n",
    "                raise Exception(\"Cannot convert from %s to %s\" % (source.__class__, self.__class__))\n",
    "\n",
    "    def make_sandwich(self):\n",
    "        \"\"\" Turns the network into a sandwich network,\n",
    "            a network with no hidden nodes and 2 layers.\n",
    "        \"\"\"\n",
    "        self.sandwich = True\n",
    "        self.cm = np.hstack((self.cm, np.zeros(self.cm.shape)))\n",
    "        self.cm = np.vstack((np.zeros(self.cm.shape), self.cm))\n",
    "        self.act = np.zeros(self.cm.shape[0])\n",
    "        return self\n",
    "        \n",
    "    def num_nodes(self):\n",
    "        return self.cm.shape[0]\n",
    "        \n",
    "    def make_feedforward(self):\n",
    "        \"\"\" Zeros out all recursive connections. \n",
    "        \"\"\"\n",
    "        if np.triu(np.nan_to_num(self.cm)).any():\n",
    "            raise Exception(\"Connection Matrix does not describe feedforward network. \\n %s\" % np.sign(self.cm))\n",
    "        self.feedforward = True\n",
    "        self.cm[np.triu_indices(self.cm.shape[0])] = 0\n",
    "        \n",
    "    def flush(self):\n",
    "        \"\"\" Reset activation values. \"\"\"\n",
    "        self.act = np.zeros(self.cm.shape[0])\n",
    "        \n",
    "    def feed(self, input_activation, add_bias=True, propagate=1):\n",
    "        \"\"\" Feed an input to the network, returns the entire\n",
    "            activation state, you need to extract the output nodes\n",
    "            manually.\n",
    "            \n",
    "            :param add_bias: Add a bias input automatically, before other inputs.\n",
    "        \"\"\"\n",
    "        if propagate != 1 and (self.feedforward or self.sandwich):\n",
    "            raise Exception(\"Feedforward and sandwich network have a fixed number of propagation steps.\")\n",
    "        act = self.act\n",
    "        node_types = self.node_types\n",
    "        cm = self.cm\n",
    "        input_shape = input_activation.shape\n",
    "        \n",
    "        if add_bias:\n",
    "            input_activation = np.hstack((1.0, input_activation))\n",
    "        \n",
    "        if input_activation.size >= act.size:\n",
    "            raise Exception(\"More input values (%s) than nodes (%s).\" % (input_activation.shape, act.shape))\n",
    "        \n",
    "        input_size = min(act.size - 1, input_activation.size)\n",
    "        node_count = act.size\n",
    "        \n",
    "        # Feed forward nets reset the activation, and activate as many\n",
    "        # times as there are nodes\n",
    "        if self.feedforward:\n",
    "            act = np.zeros(cm.shape[0])\n",
    "            propagate = len(node_types)\n",
    "        # Sandwich networks only need to activate a single time\n",
    "        if self.sandwich:\n",
    "            propagate = 1\n",
    "        for _ in range(propagate):\n",
    "            act[:input_size] = input_activation.flat[:input_size]\n",
    "            \n",
    "            if self.sum_all_node_inputs:\n",
    "                nodeinputs = np.dot(self.cm, act)\n",
    "            else:\n",
    "                nodeinputs = self.cm * act\n",
    "                nodeinputs = [ni[-np.isnan(ni)] for ni in nodeinputs]\n",
    "            \n",
    "            if self.all_nodes_same_function:\n",
    "                act = node_types[0](nodeinputs)\n",
    "            else:\n",
    "                for i in range(len(node_types)):\n",
    "                    act[i] = node_types[i](nodeinputs[i])\n",
    "\n",
    "        self.act = act\n",
    "\n",
    "        # Reshape the output to 2D if it was 2D\n",
    "        if self.sandwich:\n",
    "            return act[act.size//2:].reshape(input_shape)      \n",
    "        else:\n",
    "            return act.reshape(self.original_shape)\n",
    "\n",
    "    def cm_string(self):\n",
    "        print(\"Connectivity matrix: %s\" % (self.cm.shape,))\n",
    "        cp = self.cm.copy()\n",
    "        s = np.empty(cp.shape, dtype='a1')\n",
    "        s[cp == 0] = ' '\n",
    "        s[cp > 0] = '+'\n",
    "        s[cp < 0] = '-'\n",
    "        return '\\n'.join([''.join(l) + '|' for l in s])\n",
    "\n",
    "    \n",
    "    def visualize(self, filename, inputs=3, outputs=1):\n",
    "        \"\"\" Visualize the network, stores in file. \"\"\"\n",
    "        if self.cm.shape[0] > 50:\n",
    "            return\n",
    "        import pygraphviz as pgv\n",
    "        # Some settings\n",
    "        node_dist = 1\n",
    "        cm = self.cm.copy()\n",
    "        # Sandwich network have half input nodes.\n",
    "        if self.sandwich:\n",
    "            inputs = cm.shape[0] // 2\n",
    "            outputs = inputs\n",
    "        # Clear connections to input nodes, these arent used anyway\n",
    "\n",
    "        G = pgv.AGraph(directed=True)\n",
    "        mw = abs(cm).max()\n",
    "        for i in range(cm.shape[0]):\n",
    "            G.add_node(i)\n",
    "            t = self.node_types[i].__name__\n",
    "            G.get_node(i).attr['label'] = '%d:%s' % (i, t[:3])\n",
    "            for j in range(cm.shape[1]):\n",
    "                w = cm[i,j]\n",
    "                if abs(w) > 0.01:\n",
    "                    G.add_edge(j, i, penwidth=abs(w)/mw*4, color='blue' if w > 0 else 'red')\n",
    "        for n in range(inputs):\n",
    "            pos = (node_dist*n, 0)\n",
    "            G.get_node(n).attr['pos'] = '%s,%s!' % pos\n",
    "            G.get_node(n).attr['shape'] = 'doublecircle'\n",
    "            G.get_node(n).attr['fillcolor'] = 'steelblue'\n",
    "            G.get_node(n).attr['style'] = 'filled'\n",
    "        for i,n in enumerate(range(cm.shape[0] - outputs,cm.shape[0])):\n",
    "            pos = (node_dist*i, -node_dist * 5)\n",
    "            G.get_node(n).attr['pos'] = '%s,%s!' % pos\n",
    "            G.get_node(n).attr['shape'] = 'doublecircle'\n",
    "            G.get_node(n).attr['fillcolor'] = 'tan'\n",
    "            G.get_node(n).attr['style'] = 'filled'\n",
    "        \n",
    "        G.node_attr['shape'] = 'circle'\n",
    "        if self.sandwich: \n",
    "            # neato supports fixed node positions, so it's better for\n",
    "            # sandwich networks\n",
    "            prog = 'neato'\n",
    "        else:\n",
    "            prog = 'dot'\n",
    "        G.draw(filename, prog=prog)\n",
    "        \n",
    "    def __str__(self):\n",
    "        return 'Neuralnet with %d nodes.' % (self.act.shape[0])\n",
    "    \n",
    "\n",
    "\"\"\" Input/output relation task. Every input and output\n",
    "    is explicitly defined. XOR is an example of this task.\n",
    "\"\"\"\n",
    "\n",
    "### IMPORTS ###\n",
    "import random\n",
    "\n",
    "# Libraries\n",
    "import numpy as np\n",
    "\n",
    "# Local\n",
    "\n",
    "\n",
    "\n",
    "class XORTask(object):\n",
    "    \n",
    "    # Default XOR input/output pairs\n",
    "    INPUTS  = [(0,0), (0,1), (1,0), (1,1)]\n",
    "    OUTPUTS = [(-1,), (1,), (1,), (-1,)]\n",
    "    EPSILON = 1e-100\n",
    "    \n",
    "    def __init__(self, do_all=True):\n",
    "        self.do_all = do_all\n",
    "        self.INPUTS = np.array(self.INPUTS, dtype=float)\n",
    "        self.OUTPUTS = np.array(self.OUTPUTS, dtype=float)\n",
    "    \n",
    "    def evaluate(self, network, verbose=False):\n",
    "        if not isinstance(network, NeuralNetwork):\n",
    "            network = NeuralNetwork(network)\n",
    "        \n",
    "        network.make_feedforward()\n",
    "        \n",
    "        pairs = list(zip(self.INPUTS, self.OUTPUTS))\n",
    "        random.shuffle(pairs)\n",
    "        if not self.do_all:\n",
    "            pairs = [random.choice(pairs)]\n",
    "        rmse = 0.0\n",
    "        for (i, target) in pairs:\n",
    "            # Feed with bias\n",
    "            output = network.feed(i)\n",
    "            # Grab the output\n",
    "            output = output[-len(target):]\n",
    "            err = (target - output)\n",
    "            err[abs(err) < self.EPSILON] = 0;\n",
    "            err = (err ** 2).mean()\n",
    "            # Add error\n",
    "            if verbose:\n",
    "                print(\"%r -> %r (%.2f)\" % (i, output, err))\n",
    "            rmse += err \n",
    "\n",
    "        score = 1/(1+np.sqrt(rmse / len(pairs)))\n",
    "        return {'fitness': score}\n",
    "        \n",
    "    def solve(self, network):\n",
    "        return int(self.evaluate(network)['fitness'] > 0.9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "inputs = 3\n",
    "outputs = 4\n",
    "nonlinearities = ['tanh']\n",
    "topology = None\n",
    "feedforward = True\n",
    "max_depth = None\n",
    "max_nodes = float('inf')\n",
    "bias_as_node = False\n",
    "initial_weight_stdev = 2.0\n",
    "p_add_neuron = 0.03\n",
    "p_add_connection = 0.3\n",
    "p_mutate_weight = 0.8\n",
    "p_reset_weight = 0.1\n",
    "p_reenable_connection = 0.01\n",
    "p_disable_connection = 0.01\n",
    "p_reenable_parent=0.25\n",
    "p_mutate_bias = 0.2\n",
    "p_mutate_type = 0.2\n",
    "stdev_mutate_weight = 1.5\n",
    "stdev_mutate_bias = 0.5\n",
    "weight_range = (-50.,50.)\n",
    "\n",
    "distance_excess_weight = 1.0\n",
    "distance_disjoint_weight = 1.0\n",
    "distance_weight = 0.4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "****** Running Generation 1 ******\n",
      "Population's average fitness: 0.42802 stdev: 0.00267\n",
      "Best individual: sal oma\n",
      "Best fitness: 0.43 - #neurons: 7 - #enabled connections: 12\n",
      "Population of 100 members in 1 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal      0     100    0.43289       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 2 ******\n",
      "Population's average fitness: 0.46581 stdev: 0.02793\n",
      "Best individual: sal urindan\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 12\n",
      "Population of 100 members in 1 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal      1     100    0.49990       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 3 ******\n",
      "Population's average fitness: 0.46542 stdev: 0.02239\n",
      "Best individual: sal urindan\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 13\n",
      "Population of 100 members in 4 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal      2      25    0.49991       0\n",
      "       madda      0      25    0.44685       0\n",
      "      bronni      0      25    0.47199       0\n",
      "        bera      0      25    0.44366       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 4 ******\n",
      "Population's average fitness: 0.48717 stdev: 0.02104\n",
      "Best individual: bronni adalu\n",
      "Best fitness: 0.50 - #neurons: 8 - #enabled connections: 12\n",
      "Population of 100 members in 5 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal      3      20    0.50000       0\n",
      "       madda      1      19    0.49984       0\n",
      "      bronni      1      19    0.50000       0\n",
      "          ri      0      22    0.49995       0\n",
      "brianicekars      0      20    0.44685       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 5 ******\n",
      "Population's average fitness: 0.48781 stdev: 0.01951\n",
      "Best individual: sal urindan\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 13\n",
      "Population of 102 members in 10 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal      4      11    0.50000       0\n",
      "       madda      2      10    0.46941       1\n",
      "      bronni      2      10    0.50000       0\n",
      "          ri      1      10    0.50000       0\n",
      "brianicekars      1      10    0.49754       0\n",
      "    grhanvil      0      10    0.50000       0\n",
      "          he      0      10    0.44217       0\n",
      "        camb      0      10    0.48074       0\n",
      "         cle      0      10    0.50000       0\n",
      "          so      0      11    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 6 ******\n",
      "Population's average fitness: 0.48122 stdev: 0.03434\n",
      "Best individual: sal urindan\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 13\n",
      "Population of 99 members in 22 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal      5       4    0.50000       0\n",
      "      bronni      3       5    0.50000       0\n",
      "          ri      2       5    0.50000       0\n",
      "brianicekars      2       4    0.49754       1\n",
      "    grhanvil      1       5    0.50000       0\n",
      "          he      1       4    0.42216       1\n",
      "         cle      1       4    0.50000       1\n",
      "          so      1       4    0.50000       1\n",
      "       saffe      0       4    0.50000       0\n",
      "       crime      0       4    0.41422       0\n",
      "       islav      0       5    0.50000       0\n",
      "         kra      0       5    0.50000       0\n",
      "       anton      0       4    0.43840       0\n",
      "         gui      0       5    0.49754       0\n",
      "     pellerm      0       4    0.38169       0\n",
      "         tho      0       5    0.50000       0\n",
      "          me      0       5    0.49997       0\n",
      "       willi      0       4    0.44217       0\n",
      "       dourt      0       4    0.44217       0\n",
      "         for      0       5    0.50000       0\n",
      "          he      0       5    0.50000       0\n",
      "         krs      0       5    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 7 ******\n",
      "Population's average fitness: 0.47202 stdev: 0.03741\n",
      "Best individual: sal urindan\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 14\n",
      "Population of 103 members in 26 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal      6       4    0.50000       1\n",
      "      bronni      4       4    0.49429       1\n",
      "          ri      3       4    0.50000       0\n",
      "brianicekars      3       4    0.45227       2\n",
      "    grhanvil      2       4    0.50000       0\n",
      "          he      2       4    0.48159       0\n",
      "         cle      2       4    0.50000       0\n",
      "          so      2       4    0.50000       2\n",
      "       saffe      1       4    0.50000       0\n",
      "       crime      1       4    0.41421       1\n",
      "       islav      1       4    0.50000       0\n",
      "         kra      1       4    0.49997       1\n",
      "       anton      1       4    0.49754       0\n",
      "         gui      1       4    0.49753       1\n",
      "     pellerm      1       4    0.41427       0\n",
      "         tho      1       4    0.41747       1\n",
      "          me      1       4    0.50000       0\n",
      "       dourt      1       4    0.42216       1\n",
      "         for      1       4    0.50000       0\n",
      "          he      1       4    0.50000       0\n",
      "         krs      1       4    0.41478       1\n",
      "        nike      0       4    0.42216       0\n",
      "   haltencou      0       4    0.41422       0\n",
      "          de      0       3    0.41611       0\n",
      "       popov      0       4    0.50000       0\n",
      "        tard      0       4    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 8 ******\n",
      "Population's average fitness: 0.48531 stdev: 0.02733\n",
      "Best individual: bronni reedane\n",
      "Best fitness: 0.50 - #neurons: 8 - #enabled connections: 12\n",
      "Population of 101 members in 20 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal      7       5    0.50000       2\n",
      "      bronni      5       5    0.50000       0\n",
      "          ri      4       5    0.50000       1\n",
      "brianicekars      4       5    0.49991       0\n",
      "    grhanvil      3       5    0.50000       0\n",
      "          he      3       5    0.49977       0\n",
      "         cle      3       6    0.50000       1\n",
      "          so      3       5    0.50000       3\n",
      "       saffe      2       5    0.50000       1\n",
      "       anton      2       5    0.49992       0\n",
      "         gui      2       5    0.41745       2\n",
      "     pellerm      2       5    0.44332       0\n",
      "         tho      2       5    0.50000       0\n",
      "          me      2       5    0.47717       1\n",
      "       dourt      2       5    0.49025       0\n",
      "          he      2       5    0.50000       1\n",
      "         krs      2       5    0.41478       2\n",
      "        nike      1       5    0.46077       0\n",
      "       popov      1       5    0.50000       1\n",
      "        tard      1       5    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 9 ******\n",
      "Population's average fitness: 0.48992 stdev: 0.02303\n",
      "Best individual: grhanvil adalu\n",
      "Best fitness: 0.50 - #neurons: 8 - #enabled connections: 15\n",
      "Population of 100 members in 18 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal      8       6    0.50000       0\n",
      "      bronni      6       6    0.50000       1\n",
      "          ri      5       6    0.50000       0\n",
      "brianicekars      5       5    0.49991       1\n",
      "    grhanvil      4       6    0.50000       0\n",
      "          he      4       5    0.49977       1\n",
      "         cle      4       5    0.48374       2\n",
      "          so      4       6    0.50000       4\n",
      "       saffe      3       6    0.49987       2\n",
      "       anton      3       5    0.49999       0\n",
      "     pellerm      3       5    0.44299       1\n",
      "       dourt      3       5    0.45799       1\n",
      "          he      3       6    0.50000       0\n",
      "         krs      3       6    0.50000       0\n",
      "       popov      2       5    0.50000       0\n",
      "        tard      2       6    0.50000       1\n",
      "       sakim      0       5    0.41421       0\n",
      "         dee      0       6    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 10 ******\n",
      "Population's average fitness: 0.49773 stdev: 0.00561\n",
      "Best individual: sal reedane\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 13\n",
      "Population of 101 members in 15 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal      9       7    0.50000       1\n",
      "      bronni      7       6    0.50000       2\n",
      "          ri      6       7    0.50000       0\n",
      "brianicekars      6       7    0.49991       2\n",
      "          he      5       6    0.49977       0\n",
      "         cle      5       7    0.48374       0\n",
      "          so      5       7    0.50000       0\n",
      "       anton      4       6    0.50000       0\n",
      "          he      4       7    0.50000       1\n",
      "         krs      4       7    0.50000       1\n",
      "       popov      3       7    0.50000       0\n",
      "        tard      3       6    0.50000       2\n",
      "tistenburchf      0       7    0.48374       0\n",
      "      amerry      0       7    0.50000       0\n",
      "      errell      0       7    0.50000       0\n",
      "Solved: None\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "****** Running Generation 11 ******\n",
      "Population's average fitness: 0.49187 stdev: 0.02339\n",
      "Best individual: sal reedane\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 13\n",
      "Population of 99 members in 11 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     10       8    0.50000       2\n",
      "      bronni      8       9    0.50000       0\n",
      "          ri      7       9    0.50000       1\n",
      "brianicekars      7       9    0.48733       3\n",
      "          he      6       9    0.49977       1\n",
      "         cle      6       9    0.50000       0\n",
      "          so      6       9    0.50000       1\n",
      "       anton      5       8    0.41392       1\n",
      "         krs      5      10    0.50000       0\n",
      "       popov      4       9    0.50000       0\n",
      "        tard      4      10    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 12 ******\n",
      "Population's average fitness: 0.49098 stdev: 0.02212\n",
      "Best individual: sal reedane\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 14\n",
      "Population of 100 members in 11 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     11       8    0.50000       3\n",
      "      bronni      9      10    0.50000       1\n",
      "          ri      8      10    0.48707       2\n",
      "brianicekars      8       9    0.48733       0\n",
      "          he      7       8    0.49973       2\n",
      "         cle      7       9    0.50000       0\n",
      "          so      7       9    0.50000       0\n",
      "         krs      6      10    0.50000       1\n",
      "       popov      5       9    0.50000       1\n",
      "        tard      5      10    0.50000       1\n",
      " kenbercheck      0       8    0.41788       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 13 ******\n",
      "Population's average fitness: 0.49062 stdev: 0.02285\n",
      "Best individual: sal reedane\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 14\n",
      "Population of 99 members in 16 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     12       5    0.50000       4\n",
      "      bronni     10       5    0.50000       2\n",
      "          ri      9       6    0.50000       0\n",
      "          he      8       6    0.49999       0\n",
      "         cle      8       6    0.50000       0\n",
      "          so      8       6    0.50000       0\n",
      "         krs      7       7    0.50000       2\n",
      "       popov      6       6    0.50000       2\n",
      "        tard      6       6    0.50000       0\n",
      " kenbercheck      1       6    0.47909       0\n",
      "         bin      0       7    0.50000       0\n",
      "       reinf      0       6    0.45227       0\n",
      "       guezg      0       7    0.49973       0\n",
      "      heimen      0       7    0.50000       0\n",
      "         pec      0       7    0.50000       0\n",
      "       haris      0       6    0.41424       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 14 ******\n",
      "Population's average fitness: 0.49571 stdev: 0.01870\n",
      "Best individual: sal reedane\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 13\n",
      "Population of 100 members in 17 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     13       5    0.50000       5\n",
      "      bronni     11       5    0.50000       3\n",
      "          ri     10       5    0.49992       1\n",
      "          he      9       6    0.50000       0\n",
      "         cle      9       6    0.50000       1\n",
      "          so      9       6    0.50000       1\n",
      "         krs      8       6    0.50000       0\n",
      "       popov      7       6    0.50000       0\n",
      "        tard      7       6    0.50000       1\n",
      " kenbercheck      2       6    0.49999       0\n",
      "         bin      1       5    0.41421       1\n",
      "       reinf      1       6    0.49998       0\n",
      "      heimen      1       6    0.50000       1\n",
      "         pec      1       6    0.50000       1\n",
      "       haris      1       6    0.50000       0\n",
      "        hala      0       7    0.50000       0\n",
      "     poreffm      0       7    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 15 ******\n",
      "Population's average fitness: 0.48484 stdev: 0.03275\n",
      "Best individual: sal reedane\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 14\n",
      "Population of 102 members in 16 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     14       6    0.50000       6\n",
      "      bronni     12       6    0.50000       0\n",
      "          ri     11       6    0.50000       0\n",
      "          he     10       5    0.50000       1\n",
      "         cle     10       6    0.50000       2\n",
      "          so     10       6    0.50000       2\n",
      "         krs      9       7    0.50000       1\n",
      "       popov      8       6    0.41384       1\n",
      "        tard      8       7    0.50000       0\n",
      " kenbercheck      3       7    0.50000       0\n",
      "         bin      2       6    0.41421       2\n",
      "       reinf      2       7    0.49998       1\n",
      "      heimen      2       7    0.50000       2\n",
      "         pec      2       7    0.50000       2\n",
      "       haris      2       6    0.41421       1\n",
      "     poreffm      1       7    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 16 ******\n",
      "Population's average fitness: 0.48799 stdev: 0.02977\n",
      "Best individual: he reedane\n",
      "Best fitness: 0.50 - #neurons: 11 - #enabled connections: 20\n",
      "Population of 100 members in 14 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     15       7    0.50000       7\n",
      "      bronni     13       7    0.50000       1\n",
      "          ri     12       7    0.50000       0\n",
      "          he     11       7    0.50000       0\n",
      "         cle     11       7    0.50000       3\n",
      "          so     11       7    0.50000       3\n",
      "         krs     10       6    0.50000       2\n",
      "       popov      9       7    0.50000       0\n",
      "        tard      9       8    0.50000       1\n",
      "         bin      3       7    0.41421       3\n",
      "      heimen      3       7    0.50000       3\n",
      "         pec      3       8    0.50000       3\n",
      "       haris      3       7    0.41421       2\n",
      "     poreffm      2       8    0.50000       1\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 17 ******\n",
      "Population's average fitness: 0.48499 stdev: 0.03242\n",
      "Best individual: he reedane\n",
      "Best fitness: 0.50 - #neurons: 11 - #enabled connections: 21\n",
      "Population of 102 members in 12 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     16       8    0.50000       8\n",
      "      bronni     14       9    0.50000       2\n",
      "          ri     13       9    0.50000       0\n",
      "          he     12       8    0.50000       1\n",
      "         cle     12       9    0.50000       4\n",
      "          so     12       9    0.50000       4\n",
      "         krs     11       8    0.50000       3\n",
      "       popov     10       7    0.50000       0\n",
      "        tard     10       8    0.50000       0\n",
      "         bin      4       9    0.41421       0\n",
      "      heimen      4       9    0.50000       4\n",
      "       haris      4       9    0.41572       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 18 ******\n",
      "Population's average fitness: 0.48456 stdev: 0.03296\n",
      "Best individual: sal reedane\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 14\n",
      "Population of 100 members in 11 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     17       9    0.50000       9\n",
      "      bronni     15      10    0.50000       3\n",
      "          ri     14       9    0.50000       0\n",
      "          he     13       8    0.50000       2\n",
      "         cle     13       9    0.50000       5\n",
      "          so     13       9    0.50000       5\n",
      "         krs     12       9    0.50000       4\n",
      "       popov     11       8    0.41421       1\n",
      "        tard     11       9    0.50000       1\n",
      "         bin      5      10    0.41421       0\n",
      "      heimen      5      10    0.50000       5\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 19 ******\n",
      "Population's average fitness: 0.47856 stdev: 0.03714\n",
      "Best individual: sal reedane\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 14\n",
      "Population of 100 members in 11 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     18       9    0.50000      10\n",
      "      bronni     16       9    0.50000       4\n",
      "          ri     15       9    0.50000       1\n",
      "          he     14       9    0.50000       3\n",
      "         cle     14       9    0.50000       6\n",
      "         krs     13       9    0.50000       5\n",
      "       popov     12       8    0.41421       0\n",
      "        tard     12       8    0.41423       2\n",
      "         bin      6       9    0.41425       0\n",
      "      heimen      6      10    0.50000       6\n",
      "       bield      0      11    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 20 ******\n",
      "Population's average fitness: 0.48468 stdev: 0.03170\n",
      "Best individual: sal reedane\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 13\n",
      "Population of 100 members in 10 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     19      10    0.50000      11\n",
      "      bronni     17      10    0.50000       5\n",
      "          ri     16      10    0.50000       2\n",
      "         cle     15       9    0.50000       7\n",
      "         krs     14      10    0.50000       6\n",
      "       popov     13       9    0.41456       0\n",
      "        tard     13       9    0.50000       0\n",
      "         bin      7      10    0.42367       0\n",
      "      heimen      7      11    0.50000       7\n",
      "       bield      1      12    0.50000       1\n",
      "Solved: None\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "****** Running Generation 21 ******\n",
      "Population's average fitness: 0.49984 stdev: 0.00058\n",
      "Best individual: sal reedane\n",
      "Best fitness: 0.50 - #neurons: 7 - #enabled connections: 13\n",
      "Population of 100 members in 11 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     20       9    0.50000      12\n",
      "      bronni     18       8    0.50000       6\n",
      "          ri     17       9    0.50000       3\n",
      "         cle     16       8    0.50000       8\n",
      "         krs     15       8    0.50000       7\n",
      "       popov     14       7    0.49774       0\n",
      "         bin      8       9    0.50000       0\n",
      "      heimen      8      10    0.50000       0\n",
      "       bield      2      10    0.50000       2\n",
      "        latz      0      11    0.50000       0\n",
      "          au      0      11    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 22 ******\n",
      "Population's average fitness: 0.48628 stdev: 0.03001\n",
      "Best individual: sal reedane\n",
      "Best fitness: 0.50 - #neurons: 8 - #enabled connections: 13\n",
      "Population of 102 members in 17 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     21       6    0.50000      13\n",
      "      bronni     19       6    0.50000       7\n",
      "          ri     18       5    0.50000       4\n",
      "         cle     17       5    0.50000       9\n",
      "         krs     16       5    0.50000       8\n",
      "       popov     15       5    0.49998       0\n",
      "         bin      9       6    0.50000       0\n",
      "      heimen      9       6    0.50000       1\n",
      "       bield      3       6    0.41421       3\n",
      "        latz      1       7    0.50000       0\n",
      "          ma      0       6    0.41421       0\n",
      "    kalafoye      0       6    0.43841       0\n",
      "          cr      0       7    0.50000       0\n",
      "          sw      0       7    0.50000       0\n",
      "          ka      0       7    0.49999       0\n",
      "       eeton      0       5    0.50000       0\n",
      "      quette      0       7    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 23 ******\n",
      "Population's average fitness: 0.48471 stdev: 0.03283\n",
      "Best individual: sal reedane\n",
      "Best fitness: 0.50 - #neurons: 8 - #enabled connections: 14\n",
      "Population of 101 members in 16 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "         sal     22       6    0.50000      14\n",
      "      bronni     20       6    0.50000       8\n",
      "          ri     19       6    0.50000       5\n",
      "         cle     18       6    0.50000      10\n",
      "         krs     17       6    0.50000       9\n",
      "       popov     16       6    0.49998       1\n",
      "         bin     10       6    0.50000       0\n",
      "      heimen     10       6    0.50000       2\n",
      "       bield      4       6    0.41422       0\n",
      "        latz      2       7    0.50000       1\n",
      "          ma      1       6    0.41421       1\n",
      "    kalafoye      1       6    0.41421       1\n",
      "          cr      1       7    0.50000       1\n",
      "          sw      1       7    0.50000       1\n",
      "       eeton      1       7    0.50000       0\n",
      "      quette      1       7    0.50000       1\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 24 ******\n",
      "Population's average fitness: 0.48970 stdev: 0.02788\n",
      "Best individual: bin reedane\n",
      "Best fitness: 0.50 - #neurons: 10 - #enabled connections: 17\n",
      "Population of 100 members in 15 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "      bronni     21       6    0.50000       9\n",
      "          ri     20       6    0.50000       6\n",
      "         cle     19       6    0.50000      11\n",
      "         krs     18       6    0.50000      10\n",
      "       popov     17       6    0.49998       2\n",
      "         bin     11       6    0.50000       0\n",
      "      heimen     11       6    0.50000       3\n",
      "       bield      5       7    0.50000       0\n",
      "        latz      3       8    0.50000       0\n",
      "          ma      2       6    0.41421       2\n",
      "    kalafoye      2       6    0.41421       2\n",
      "          cr      2       8    0.50000       2\n",
      "          sw      2       8    0.50000       2\n",
      "       eeton      2       7    0.50000       1\n",
      "    schaelsd      0       8    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 25 ******\n",
      "Population's average fitness: 0.49327 stdev: 0.02306\n",
      "Best individual: bin reedane\n",
      "Best fitness: 0.50 - #neurons: 10 - #enabled connections: 17\n",
      "Population of 102 members in 13 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "      bronni     22       7    0.50000      10\n",
      "          ri     21       7    0.50000       7\n",
      "         cle     20       8    0.50000      12\n",
      "         krs     19       8    0.50000      11\n",
      "       popov     18       7    0.50000       0\n",
      "         bin     12       7    0.50000       1\n",
      "      heimen     12       7    0.50000       4\n",
      "       bield      6       8    0.50000       1\n",
      "          ma      3       8    0.50000       0\n",
      "    kalafoye      3       8    0.41421       3\n",
      "          cr      3       9    0.50000       3\n",
      "          sw      3       9    0.50000       3\n",
      "    schaelsd      1       9    0.50000       1\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 26 ******\n",
      "Population's average fitness: 0.50469 stdev: 0.01720\n",
      "Best individual: bin reedane\n",
      "Best fitness: 0.57 - #neurons: 10 - #enabled connections: 17\n",
      "Population of 101 members in 13 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "      bronni     23       7    0.50000      11\n",
      "          ri     22       7    0.50000       8\n",
      "         cle     21       7    0.50000      13\n",
      "         krs     20       7    0.50000      12\n",
      "       popov     19       7    0.50000       0\n",
      "         bin     13       7    0.56771       0\n",
      "      heimen     13       7    0.50000       5\n",
      "       bield      7       9    0.50000       2\n",
      "          ma      4       8    0.50000       0\n",
      "          cr      4       8    0.50000       4\n",
      "          sw      4       9    0.50000       4\n",
      "    schaelsd      2       9    0.50000       2\n",
      "    unneerme      0       9    0.50000       0\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 27 ******\n",
      "Population's average fitness: 0.50000 stdev: 0.00001\n",
      "Best individual: bin reedane\n",
      "Best fitness: 0.50 - #neurons: 11 - #enabled connections: 18\n",
      "Population of 99 members in 11 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "      bronni     24       8    0.50000      12\n",
      "          ri     23       9    0.50000       0\n",
      "         cle     22       8    0.50000      14\n",
      "         krs     21       8    0.50000      13\n",
      "       popov     20       8    0.50000       0\n",
      "         bin     14       8    0.50002       1\n",
      "       bield      8       9    0.50000       0\n",
      "          ma      5      11    0.50000       1\n",
      "          cr      5      10    0.50000       5\n",
      "          sw      5      10    0.50000       5\n",
      "    schaelsd      3      10    0.50000       3\n",
      "Solved: None\n",
      "\n",
      "****** Running Generation 28 ******\n",
      "Population's average fitness: 0.52565 stdev: 0.11722\n",
      "Best individual: ri reedane\n",
      "Best fitness: 0.91 - #neurons: 10 - #enabled connections: 17\n",
      "Population of 99 members in 11 species:\n",
      "Species         age    size    fitness    stag\n",
      "============    ===    ====    =======    ====\n",
      "      bronni     25       8    0.50000      13\n",
      "          ri     24       8    0.91248       0\n",
      "         krs     22       8    0.50000      14\n",
      "       popov     21       8    0.50000       0\n",
      "       bield      9       9    0.50000       1\n",
      "          ma      6       9    0.50000       2\n",
      "          cr      6      10    0.50000       6\n",
      "          sw      6      10    0.50000       6\n",
      "    schaelsd      4      10    0.50000       4\n",
      "          pa      0      10    0.50000       0\n",
      "        tope      0       9    0.41553       0\n",
      "Solved: 28\n"
     ]
    }
   ],
   "source": [
    "population_size = 100\n",
    "elitism = False\n",
    "stop_when_solved = True \n",
    "tournament_selection_k = 3 \n",
    "verbose = True\n",
    "max_cores = 1\n",
    "compatibility_threshold = 3.0\n",
    "compatibility_threshold_delta = 0.4 \n",
    "target_species = 12\n",
    "minimum_elitism_size = 5 \n",
    "young_age = 10\n",
    "young_multiplier = 1.2 \n",
    "old_age = 30\n",
    "old_multiplier = 0.2 \n",
    "stagnation_age = 15\n",
    "reset_innovations = False\n",
    "survival = 0.2\n",
    "\n",
    "genome_factory = lambda: Genotype(inputs, outputs, nonlinearities, topology, feedforward, max_depth, max_nodes, initial_weight_stdev, bias_as_node, p_add_neuron, p_add_connection, p_mutate_weight, p_reset_weight, p_reenable_connection, p_disable_connection, p_reenable_parent, p_mutate_bias, p_mutate_type, stdev_mutate_weight, stdev_mutate_bias, weight_range, distance_excess_weight, distance_disjoint_weight, distance_weight)\n",
    "\n",
    "population = Population(genome_factory, population_size, elitism, stop_when_solved, tournament_selection_k, verbose, max_cores, compatibility_threshold, compatibility_threshold_delta, target_species, minimum_elitism_size, young_age, young_multiplier, old_age, old_multiplier, stagnation_age, reset_innovations, survival)\n",
    "task = XORTask()\n",
    "\n",
    "result = population.epoch(evaluator = task, generations = 1000, solution = task)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[-0.1557,  0.6790,  1.2225, -0.5605, -0.0323],\n",
       "        [-0.6397, -0.1951,  0.3126,  0.9808, -0.4102],\n",
       "        [-0.8842,  1.1706, -0.5982,  0.6167,  0.3099]])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "w = torch.empty(3, 5)\n",
    "nn.init.xavier_uniform_(w, gain=nn.init.calculate_gain('relu'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
